android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -146,40 +147,54 @@ class TranslationsMiddleware(
-     * [TranslationsAction.InitTranslationsBrowserState].
+     * Use this to initialize translations data for [BrowserState.translationEngine]. If an
+     * issue occurs, the relevant error will be set on [BrowserState.translationEngine].
+     *
+     * This will populate:
+     * Language Support - [requestSupportedLanguages]
      *
      * @param context Context to use to dispatch to the store.
      */
-    private fun requestEngineSupportAndInit(
+    private fun initializeBrowserStore(
         context: MiddlewareContext<BrowserState, BrowserAction>,
     ) {
-        engine.isTranslationsEngineSupported(
-            onSuccess = { isEngineSupported ->
-                if (isEngineSupported) {
+        requestSupportedLanguages(context)
+    }
+
+    /**
+     * Checks if the translations engine supports the device architecture and updates the state on
+     * [BrowserState.translationEngine].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @return Whether the engine is supported or not, or null when the support cannot be
+     * determined.
+     */
+    private suspend fun requestEngineSupport(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+    ): Boolean? {
+        return suspendCoroutine { continuation ->
+            engine.isTranslationsEngineSupported(
+                onSuccess = { isEngineSupported ->
                     context.store.dispatch(
-                        TranslationsAction.InitTranslationsBrowserState,
+                        TranslationsAction.SetEngineSupportedAction(
+                            isEngineSupported = isEngineSupported,
+                        ),
                     )
-                }
-                context.store.dispatch(
-                    TranslationsAction.SetEngineSupportedAction(
-                        isEngineSupported = isEngineSupported,
-                    ),
-                )
-                logger.info("Success requesting engine support.")
-            },
+                    logger.info("Success requesting engine support.")
+                    continuation.resume(isEngineSupported)
+                },
 
-            onError = { error ->
-                context.store.dispatch(
-                    TranslationsAction.EngineExceptionAction(
-                        error = TranslationError.UnknownEngineSupportError(error),
-                    ),
-                )
-                logger.error("Error requesting engine support: ", error)
-            },
-        )
+                onError = { error ->
+                    context.store.dispatch(
+                        TranslationsAction.EngineExceptionAction(
+                            error = TranslationError.UnknownEngineSupportError(error),
+                        ),
+                    )
+                    logger.error("Error requesting engine support: ", error)
+                    continuation.resume(null)
```

Comment:
I see this is landing already and I think we could handle any updates to this in a follow-up, but wondering if this could just be `false` instead of `null`? Is there a functional difference between the false case above and the exception case here?

Reply:
Good thing to mention and thanks for taking a look! It is following a similar pattern as these [functions](https://searchfox.org/mozilla-mobile/rev/484082df2b810fe8ea7ec2167e63e063f907b0a7/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt#391,411) for parallelism. I think keeping the `true = yes, support` / `false = no support` / `null = unknown support` state distinction is helpful since we might want to retry on the unknown or null case. (Still thinking through how retry logic is going to work all around, but the distinction will show back up in some form.)

Reply:
(Happened to notice in those linked functions that they are missing logging - added that to my list of things to refactor when in the area!)